### PR TITLE
Separate AnonConst from ConstBlock in HIR.

### PR DESCRIPTION
**Pull request by cjgillot**
Their behaviours are different enough to justify having separate nodes.

### CODE REVIEW

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -465,6 +465,17 @@ impl<'tcx> Visitor<'tcx> for MarkSymbolVisitor<'tcx> {
 
         self.in_pat = in_pat;
     }
+
+    fn visit_inline_const(&mut self, c: &'tcx hir::ConstBlock) {
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/dead.rs:469]: Do we still need the above `visit_anon_const` if const blocks should be handled in `visit_inline_const` instead?

Comment (cjgillot) [reply compiler/rustc_passes/src/dead.rs:470]: Yes. You could still want to mark `three` as live in:
```rust
match S::<3> {
    S::<{three()}> => {}
}
```

**Comment by BoxyUwU** [comment]
What are the differing behaviours of anon consts and inline const that motivate this? Off the top of my head inline consts get typeck'd with their parent item and have weirder generics/typeof to allow for that

**Comment by apiraino** [comment]
switching to author to reply to [previous comment](https://github.com/rust-lang/rust/pull/109609#issuecomment-1525474614) and rebase - thanks :)

@rustbot author

**Comment by cjgillot** [comment]
> What are the differing behaviours of anon consts and inline const that motivate this? Off the top of my head inline consts get typeck'd with their parent item and have weirder generics/typeof to allow for that wheras other anon consts get typeckd by themselves thinking

Yes, the difference in typeck root is the main reason to separate those cases IMO. Most of the cases where we end-up with code duplication are in side passes, with much less complexity than type collection.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 2]
```diff
@@ -1189,7 +1189,7 @@ fn should_encode_type(tcx: TyCtxt<'_>, def_id: LocalDefId, def_kind: DefKind) ->
 
 fn should_encode_const(def_kind: DefKind) -> bool {
     match def_kind {
-        DefKind::Const | DefKind::AssocConst | DefKind::AnonConst => true,
+        DefKind::Const | DefKind::AssocConst | DefKind::AnonConst | DefKind::InlineConst => true,
```
Comment (BoxyUwU) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1192]: Is this change needed because of this PR or was this an existing bug in `feature(inline_const)`?

**Review by BoxyUwU — APPROVED** [review]
sorry for taking so long to get to looking at this properly. r=me after rebasing

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1194]: This was an existing bug.
